syntax = "proto3";

package im;

import "common.proto";

service SessionService {
    rpc CreateSession(CreateSessionRequest) returns (CreateSessionResponse);
    rpc DeleteSession(DeleteSessionRequest) returns (DeleteSessionResponse);
    rpc GetSessions(GetSessionsRequest) returns (GetSessionsResponse);
    rpc GetSessionInfo(GetSessionInfoRequest) returns (GetSessionInfoResponse);
    rpc AddFriend(AddFriendRequest) returns (AddFriendResponse);
    rpc RemoveFriend(RemoveFriendRequest) returns (RemoveFriendResponse);
    rpc GetFriends(GetFriendsRequest) returns (GetFriendsResponse);
    rpc GetFriendDetail(GetFriendDetailRequest) returns (GetFriendDetailResponse);
    rpc SearchFriends(SearchFriendsRequest) returns (SearchFriendsResponse);
}

message CreateSessionRequest {
    int64 user_id = 1;
    int64 peer_id = 2;
}

message CreateSessionResponse {
    bool success = 1;
    Error error = 2;
    int64 session_id = 3;
}

message DeleteSessionRequest {
    int64 user_id = 1;
    int64 peer_id = 2;
}

message DeleteSessionResponse {
    bool success = 1;
    Error error = 2;
}

message GetSessionsRequest {
    int64 user_id = 1;
    int32 limit = 2;
    int64 offset = 3;
}

message GetSessionsResponse {
    bool success = 1;
    Error error = 2;
    repeated SessionInfo sessions = 3;
    int32 total_count = 4;
}

message GetSessionInfoRequest {
    int64 session_id = 1;
}

message GetSessionInfoResponse {
    bool success = 1;
    Error error = 2;
    SessionInfo session = 3;
}

message AddFriendRequest {
    int64 user_id = 1;
    int64 friend_id = 2;
    string remark = 3;
    string message = 4;
}

message AddFriendResponse {
    bool success = 1;
    Error error = 2;
    int64 friend_id = 3;
}

message RemoveFriendRequest {
    int64 user_id = 1;
    int64 friend_id = 2;
}

message RemoveFriendResponse {
    bool success = 1;
    Error error = 2;
}

message GetFriendsRequest {
    int64 user_id = 1;
    int32 limit = 2;
    int64 offset = 3;
}

message GetFriendsResponse {
    bool success = 1;
    Error error = 2;
    repeated FriendInfo friends = 3;
    int32 total_count = 4;
}

message GetFriendDetailRequest {
    int64 user_id = 1;
    int64 friend_id = 2;
}

message GetFriendDetailResponse {
    bool success = 1;
    Error error = 2;
    FriendInfo friend = 3;
}

message SearchFriendsRequest {
    int64 user_id = 1;
    string keyword = 2;
}

message SearchFriendsResponse {
    bool success = 1;
    Error error = 2;
    repeated UserInfo users = 3;
}

message SessionInfo {
    int64 session_id = 1;
    int64 peer_id = 2;
    SessionType session_type = 3;
    string peer_name = 4;
    string peer_avatar = 5;
    string last_message = 6;
    int64 last_message_time = 7;
    int32 unread_count = 8;
}

message FriendInfo {
    int64 user_id = 1;
    string nickname = 2;
    string avatar = 3;
    string remark = 4;
    OnlineStatus status = 5;
    int64 last_active = 6;
}

message UserInfo {
    int64 user_id = 1;
    string nickname = 2;
    string avatar = 3;
    OnlineStatus status = 4;
}
